Restoration Druid Rework

syrifgit opened the pull request:
# Restoration Druid TWW Modelling Cleanup

## Guardian Cleanup
- Account for `Master Shapeshifter` talent mana refund for Resto Druids when casting `Ironfur`

## Feral Cleanup
- Account for `Master Shapeshifter` talent mana refund for Resto Druids when casting `Rip`, `Ferocious Bite` or `Maim` with 5 combo points

## Balance Cleanup
- `Nature's Swiftness` stuff
- `Rebirth` to account for buff consumptions
- Separate `Starsurge` into a specific Balance and Resto version

## Restoration Cleanup

### Basics
- `strformat`
- Remove `Eclipse` stuff
- More `Grove Guardians` NPC support
- Default settings
- Pack string
- Shapeshift form logic tested and corrected

### Spell, Buff, Talent Remodelling
- `Tranquility`
  - Proper support for HoT stack application and CDR emulation with `Dreamstate` talent 
- `Incarnation` / `Tree of Life`
  - All SpellIDs corrected, behaviour modelled correctly, `Grove Guardians` summon emulation
- `Cenarion Ward`
  - Shield and HoT tracking implemented via `active_dot.cenarion_ward`
- `Efflorescence`
  - Updated buff tracking for use with APL via `buff.efflorescence`
  - Spring Blossoms HoT tracking implemented via `active_dot.spring_blossoms`
- `Flourish`
  - Active state tracking for use with APL via `buff.flourish`
- `Grove Guardians`
  - Correct Cooldown, tracking buff and `summonPet`
- `Incarnation`
  - Spell and buffs remodelled to reflect correct SpellIDs, form shifting, and buffs to the empowered spells during shapeshift form
  - Support for `Grove Guardian` summon emulation with the `Cenarius' Guidance` talent
- `Ironbark`
  - Cooldown now accounts for `Improved Ironbark` talent
- `Lifebloom`
  - HoT Application tracking implemented via `active_dot.lifebloom`
  - SpellID corrected (it changes with `Undergrowth` talent
- `Nature's Swiftness`
  - Cooldown accounts for `Passing Seasons` talent
- `Nourish`
  - Cast speed modelling, including `Wild Synthesis` talent / buff
- Regrowth
  - Cast speed model updated to reflect current iteration of `Clearcasting` and `Incarnation`
  - Cost model updated to reflect `Clearcasting` and `Abundance`
  - Buff consumption
  - HoT application tracking implemented via `active_dot.regrowth`
- `Rejuvenation`
  - Cost model updated to reflect only being **_in tree form_** during incarnation
  - HoT application tracking implemented via `active_dot.rejuvenation`
  - `Germination` HoT application tracking implemented via `active_dot.germination`
  - `Cultivation` modelling added, /w HoT application tracking implemented via `active_dot.cultivation`
- `Soul of the Forest` talent
  - Support added in all relevant spells + aura
- `Swiftmend`
  - Charge/Recharge converted to `if then return end end` format
  - `Reforestation` modelling /w aura and `Tree of Life` shifting
- Wild Growth
  - HoT application tracking implemented via `active_dot.wild_growth`, and incorporates talents
- `Heart of the Wild`
  - Support for `Call of the Archdruid `talent
  - Emulation of periodic combo point gains
- `Starsurge`, `Starfire` and `Wrath`
  - Remodelled as their own unique versions compared to Balance, support for any and all shapeshifting, buffs, mana returns, costs, cast times
### Spec Options
Healing mode, similar to Restoration Shaman. Opt-In.
- Maintain Efflorescence
- Maintain atleast 1 rejuvenation (for the tank)
- Maintain Lifebloom (1 or 2 depending on your talents)
- Always wild growth after a swiftmend
- Tree-proc swiftmends

syrifgit committed:
Initial Resto Druid pass

Done
Rejuv
Regrowth
Efflo
Lifebloom
Flourish
Tranq
Invig
Wrath
Incarn
Swiftmend

TODO
Starsurge/starfire/HoTW
Catform
Bearform
Natures swiftness
wild growth active dots
grove guardians
wildstalker

syrifgit committed:
More modelling

syrifgit committed:
More modeling

syrifgit committed:
Spec options and actionlist

syrifgit committed:
Final proofread + formatting

Hekili commented on TheWarWithin/DruidBalance.lua:
```diff
@@ -2874,7 +2888,7 @@ spec:RegisterAbilities( {
 
 
     starsurge = {
-        id = function() return state.spec.balance and 78674 or 197626 end,
```
This is the only version of Starsurge that's in the Druid files; theoretically, other specs can use Starsurge under come conditions, no?

syrifgit replied:
I'll check, it may be that I need to bump the resto version up into feral, or just copy it to all 3 other specs. The specs go alphabetically by filename when trying to call registrations from other specs, right?

Hekili replied:
I think it goes in load order, with current spec superceding.  Which I think tends to be alphabetical order.

syrifgit replied:
I've copied it onto Guardian and Feral, their versions most closely match the Resto version, and the talent checks take care of the rest anyways. And they both use the resto spell ID, Balance is the odd one out.

Hekili commented on TheWarWithin/DruidRestoration.lua:
```diff
@@ -202,20 +203,24 @@ spec:RegisterAuras( {
     },
     call_of_the_elder_druid = {
-        id = 338643,
+        id = 3426790,
```
This is a typo.

syrifgit replied:
`426790`

Hekili commented on TheWarWithin/DruidRestoration.lua:
```diff
@@ -276,40 +284,50 @@ spec:RegisterAuras( {
+    -- The actual incarn buff
+    incarnation = {
+        id = 117679,
+        duration = 3600,
```
Fix duration.

syrifgit replied:
Had them backwards

Hekili commented on TheWarWithin/DruidRestoration.lua:
```diff
@@ -276,40 +284,50 @@ spec:RegisterAuras( {
+    -- This is the form
     incarnation_tree_of_life = {
         id = 33891,
         duration = 30,
```
Fix duration.  incarnation and incarnation_tree_of_life should be same aura; this can be tree_of_life / tree_of_life_form.

syrifgit replied:
They are distinct in-game and the registration name is their real name in-game.

```
player_buffs:
     1126 - mark_of_the_wild                         -   1 - 2446.09
    33891 - incarnation_tree_of_life                 -   1 - 3600.00
...
...
...
...
   117679 - incarnation                              -   1 - 28.64 
```

Do you still want to rename the form one? or just leave as copy name for use in the code/APL?

![Screenshot 2024-11-11 215646](https://github.com/user-attachments/assets/341ccf09-1f3f-4133-a81d-daadca0b3af1)


![image](https://github.com/user-attachments/assets/aab4a128-6bfa-484e-9d51-35aa7d4403e4)

Hekili replied:
No, I didn't realize the aura design in-game was dumb.  The critical piece is that the tokens match SimC and any priority logic.

Hekili reviewed (changes requested):
Is Starsurge usable in Feral or Guardian ever?  (Heart of the Wild?)  I just want to be sure that the right Starsurge is available for them since the Balance version will probably supercede the Resto version for those specs.

syrifgit committed:
Review Notes

Still need decision on tree of life name

syrifgit committed:
hasteCDs and charge functions

Coding convention consistency

syrifgit committed:
Merge branch 'Hekili:thewarwithin' into syrif-druid

syrifgit committed:
Resto Druid Final Copy

Don't overwrite your entire PR with an old version this time.

Fixed all `cast`, `cooldown` and `charge / recharge`. Review notes implemented.

syrifgit commented:
Actually ready now.
